### PR TITLE
enh(api) AddHostTemplate: handle macros

### PR DESCRIPTION
**Pull request by a-launois**
## Description

Update AddHostTemplate to handle macros

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 1]
```diff
@@ -407,4 +236,119 @@ private function findParentTemplates(array $templateIds): array
...
+            );
+
+            $commandMacros = MacroManager::resolveInheritanceForCommandMacro($existingCommandMacros);
+        }
+
+        return [$inheritedMacros, $commandMacros];
+    }
+
+    /**
+     * @param int $hostTemplateId
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:325]: Missing phpdoc explaining there is no ihnrtied macro in the response by purpose.

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 2]
```diff
@@ -407,4 +236,119 @@ private function findParentTemplates(array $templateIds): array
...
+        $inheritanceLine = HostInheritance::findInheritanceLine($hostTemplateId, $templateParents);
+        $existingHostMacros = $this->readHostMacroRepository->findByHostIds($inheritanceLine);
+
+        [$directMacros, $inheritedMacros] = MacroManager::resolveInheritanceForHostMacro(
+            $existingHostMacros,
+            $inheritanceLine,
+            $hostTemplateId
+        );
+
+        /** @var array<string,CommandMacro> */
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:310]: Don't forget variable, official syntax is `/** @var MyType $variable */`
```suggestion
        /** @var array<string,CommandMacro> $commandMacros */
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 3]
```diff
@@ -407,4 +236,119 @@ private function findParentTemplates(array $templateIds): array
...
+        }
+
+    }
+
+    /**
+     * Find macros of a host template:
+     * macros linked through template inheritance, macros linked through command inheritance.
+     *
+     * @param int $hostTemplateId
+     * @param int $checkCommandId
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:289]:
```suggestion
     * @param ?int $checkCommandId
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        private readonly ContactInterface $user
+    ) {
+    }
+
+    /**
+     * Assert name is not already used.
+     *
+     * @param string $name
+     *
+     * @throws HostTemplateException
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php:62]:
```suggestion
     * @throws HostTemplateException|\Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    /**
+     * Assert time period ID is valid.
+     *
+     * @param ?int $timePeriodId
+     * @param ?string $propertyName
+     *
+     * @throws HostTemplateException
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php:96]:
```suggestion
     * @throws HostTemplateException|\Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+            throw HostTemplateException::nameAlreadyExists($formattedName, $name);
+        }
+    }
+
+    /**
+     * Assert icon ID is valid.
+     *
+     * @param ?int $iconId
+     *
+     * @throws HostTemplateException
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php:79]:
```suggestion
     * @throws HostTemplateException|\Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+     *
+     * @param string $name
+     *
+     * @throws HostTemplateException
+     */
+    public function assertIsValidName(string $name): void
+    {
+        $formattedName = HostTemplate::formatName($name);
+        if ($this->readHostTemplateRepository->existsByName($formattedName)) {
+            $this->error('Host template name already exists', ['name' => $name, 'formattedName' => $formattedName]);
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplateValidation.php:68]: Just for knowledge about [compact](https://www.php.net/manual/en/function.compact.php) 😉 
💯 equivalent in this case.
```suggestion
            $this->error('Host template name already exists', compact('name', 'formattedName'));
```

**Review by abuathier — DISMISSED** [review]

**Review by dmyios — APPROVED** [review]

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php (hunk 1 above)
Comment (a-launois) [reply centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:325]: It is stipulated in the API doc

**Review by abuathier — APPROVED** [review]

**Pull request merged**